staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go

```diff
@@ -225,38 +304,79 @@ func Start(nodeServer interface{}, opts ...Option) (result DRAPlugin, finalErr e
 		return nil, fmt.Errorf("start node client: %v", err)
 	}
 	d.plugin = plugin
-	defer func() {
-		// Clean up if we didn't finish succcessfully.
-		if r := recover(); r != nil {
-			plugin.stop()
-			panic(r)
-		}
-		if finalErr != nil {
-			plugin.stop()
-		}
-	}()
 	if !implemented {
 		return nil, errors.New("no supported DRA gRPC API is implemented and enabled")
 	}
 
 	// Now make it available to kubelet.
-	registrar, err := startRegistrar(klog.LoggerWithName(o.logger, "registrar"), o.grpcVerbosity, o.unaryInterceptors, o.streamInterceptors, o.driverName, o.draAddress, o.pluginRegistrationEndpoint)
+	registrar, err := startRegistrar(klog.NewContext(ctx, klog.LoggerWithName(logger, "registrar")), o.grpcVerbosity, o.unaryInterceptors, o.streamInterceptors, o.driverName, o.draAddress, o.pluginRegistrationEndpoint)
 	if err != nil {
 		return nil, fmt.Errorf("start registrar: %v", err)
 	}
 	d.registrar = registrar
 
+	// startGRPCServer and startRegistrar don't implement cancellation
+	// themselves, we add that for both here.
+	d.wg.Add(1)
+	go func() {
+		defer d.wg.Done()
+		<-ctx.Done()
+		d.plugin.stop()
+		d.registrar.stop()
+	}()
+
 	return d, nil
 }
 
+// Stop implements [DRAPlugin.Stop].
 func (d *draPlugin) Stop() {
 	if d == nil {
 		return
 	}
+	d.cancel(errors.New("DRA plugin was stopped"))
 	d.registrar.stop()
 	d.plugin.stop()
+
+	// d.resourceSliceController is set concurrently.
+	// Stop may get called with nil receiver, so we
+	// don't need to check here.
+	d.mutex.Lock()
+	d.resourceSliceController.Stop()
+	d.mutex.Unlock()
+
+	d.wg.Wait()
+}
+
+// PublishResources implements [DRAPlugin.PublishResources].
+func (d *draPlugin) PublishResources(ctx context.Context, nodeResources []*resourceapi.ResourceModel) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	owner := resourceslice.Owner{
+		APIVersion: "v1",
+		Kind:       "Node",
+		Name:       d.nodeName,
+		// UID will be determined by the controller.
```

Comment:
I still think we should provide a way for the caller to supply this if they already have access to it. I'm fine making this optional, but it seems unnecessary to make an extra APIserver call to retrieve this if we've already done this once in the driver.

Reply:
Or alternatively (as mentioned in a comment below), can we make the APIserver call here instead of in the controller? The part I motly dislike isn't the APIserver call itself, it's passing an incomplete object like this, assuming it will then be filled out with the rest of the information later.

Reply:
> I'm fine making this optional, but it seems unnecessary to make an extra APIserver call to retrieve this if we've already done this once in the driver.

But typically the driver will not have done that, will it? It gets the node name through the downward API and then just passes that on.

> Can we make the APIserver call here instead of in the controller?

That implies adding another retry loop and blocking in `PublishResources` until the UID has been retrieved. That can fail, which means adding an error return value to the method and then somehow handling that failure in the caller. Is that really better?

Reply:
I added a `NodeUID` option, although I don't think that it'll be used much. See https://github.com/kubernetes/kubernetes/compare/36b913f2ed61a881e764e592f5e7b18984fca640..08376256a05bbd9918311cb9ad2ebee35a95c25f